dm3-names/contracts/l2/Dm3NameRegistrar.sol

```diff
@@ -0,0 +1,149 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.17;
+
+import {IAddrResolver} from '@ensdomains/ens-contracts/contracts/resolvers/profiles/IAddrResolver.sol';
+import {INameResolver} from '@ensdomains/ens-contracts/contracts/resolvers/profiles/INameResolver.sol';
+import {ITextResolver} from '@ensdomains/ens-contracts/contracts/resolvers/profiles/ITextResolver.sol';
+
+/// @title Dm3NameRegistrar
+/// @notice This contract is used for registering names in the ENS system. It is a combination of ENSResolver and ReverseRegistrar contracts. Allowing to register names and set text records for each name. By beeing compatible with ENSResolver and ReverseRegistrar
+
+contract Dm3NameRegistrar is IAddrResolver, INameResolver, ITextResolver {
+    //Lookup table for hexadecimal conversion
+    //Taken from ENS ReverseRegistrar contract
+    //https://github.com/ensdomains/ens-contracts/blob/21736916300b26cb8ea1802dbf6c9ff054adaeab/contracts/reverseRegistrar/ReverseRegistrar.sol#L12
+    bytes32 private constant lookup =
+        0x3031323334353637383961626364656600000000000000000000000000000000;
+
+    // Constant for reverse node address
+    //Taken from ENS ReverseRegistrar contract
+    //https://github.com/ensdomains/ens-contracts/blob/21736916300b26cb8ea1802dbf6c9ff054adaeab/contracts/reverseRegistrar/ReverseRegistrar.sol#L12
+    bytes32 private constant ADDR_REVERSE_NODE =
+        0x91d1777781884d03a6757a803996e38de2a42967fb37eeaca72729271025a9e2;
+
+    //Node of the L1 domain. in case of OP name that would be namehash(op.dm3.eth)
+    //Is not immutable so it can be retrieved from the storage using CCIP
+    bytes32 public PARENT_NODE;
+
+    // Mapping to store the owner of each node
+    mapping(bytes32 => address) public owner;
+
+    // Mapping to store the reverse record of each node
+    mapping(bytes32 => string) public reverse;
+
+    // Mapping to store text records for each node
+    mapping(bytes32 => mapping(string => string)) public texts;
+
+    // Event emitted when a name is registered
+    event NameRegistered(address indexed addr, string indexed name);
+
+    // Event emitted when a name is removed
+    event NameRemoved(address indexed addr, string indexed name);
+
+    /// @notice Constructor to set the parent node
+    /// @param _parentNode The parent node of the ENS
+    constructor(bytes32 _parentNode) {
+        PARENT_NODE = _parentNode;
+    }
+
+    /// @notice Register a name in the ENS system
+    /// @param _name The name to register
+    function register(string calldata _name) external {
+        string memory oldName = reverse[makeReverseNode(msg.sender)];
+        if (bytes(_name).length == 0) {
+            // Clear name if the new name is empty
+            delete owner[makeLabelNode(oldName)];
+            delete reverse[makeReverseNode(msg.sender)];
```

Comment:
deleting test records seems a good idea, too.

Reply:
Todo add implement resolverBase so text records can be versioned